### ISSUE TITLE
StopIteration

### ISSUE THREAD

**Issue by angelaparodymerino** [harvardinformatics/TranscriptomeAssemblyTools#3]
Hi,

I am trying to get a transcriptome from some SRA data of 24 samples, same species. I have used Rcorrector which generated .cor.fq (R1 and R2) files of 200Kb-2.5Gb. Now want to run FilterUncorrectabledPEfastq.py following Best Practices for de novo Transcriptome Assembly (https://informatics.fas.harvard.edu/best-practices-for-de-novo-transcriptome-assembly-with-trinity.html) since this is new to me. Some of my samples run without errors and produced three files:

unfixrm_samplename_1.cor.fq
unfixrm_samplename_2.cor.fq
rmumfixable_samplename.log

One example of .log file:

total PE reads:2134565
removed PE reads:174989
retained PE reads:1959576
R1 corrected:614746
R2 corrected:695033
pairs corrected:978537
R1 unfixable:65169
R2 unfixable:90032
both reads unfixable:19788

Samples that seem not to have run well did not produce .log file but did produce unfixrm.....cor.fq files. This is the end of the run of one of those samples (see error message at the end).

14200000 reads processed
14300000 reads processed
14400000 reads processed
14500000 reads processed
14600000 reads processed
Traceback (most recent call last):
  **File "FilterUncorrectabledPEfastq.py", line 78, in <module>
    head2,seq2,placeholder2,qual2=[j.strip() for j in R2.next()]
StopIteration**

1) What does this message mean? and what do yo suggest me to do?
2) Just to be sure, unfixrm...1.cor.fq and unfixrm...2.cor.fq are the new and fixed fq files for next step of processing (which will be trim adapter trim by quality).

Let me know if you need more details.

Any other advice, further explanation or suggestion is welcome, since I am new procesing RNAseq data and always happy to learn more!

Thanks in advance

Cheers,

'Angela

**Comment by adamfreedman** [comment]
Hi Angela,

The reason why the unfixrm*fastq files are written is because file handles for those outputs are opened earlier in the script ... and ... the script writes as it processes reads from the original input files. Because you encountered an error partway, those files are truncated. Furthermore, no logs are written because the log file doesn't get written until all of the reads have been processed.

The error you are encountering occurs when an R2 read is being read from the input stream such that it can be paired with the matching R2 file. My guess is that, somehow, you have different numbers of R1 and R2 reads such that you go looking for the next R2 read, and it isn't there. That would explain the "stop iteration". Do a quick line count on the R1 and R2 files and see if they are the same. For all you know, the researchers who generated the files and posted them to SRA screwed up the upload, or uploaded processed files where they had selective removed reads, but not in a consistent way so that the same number of reads were retained for R1 and R2.

**Comment by angelaparodymerino** [comment]
Thanks for the quick answer.

I have checked the initial fasq files (the input files used for Rcorrector) and R1 and R2 of each sample have the same number of reads. However, the outputs of Rcorrector (samplename_1.cor.fq, samplename_2.cor.fq) of each sample contain different number of reads. 


~/data/RNAseqCorrected$ wc -l SRR4030252_1.cor.fq
**47012927** SRR4030252_1.cor.fq
~/data/RNAseqCorrected$ wc -l SRR4030252_2.cor.fq
**47012931** SRR4030252_2.cor.fq

~/data/RNAseqCorrected$ wc -l SRR4030253_1.cor.fq
**31358341** SRR4030253_1.cor.fq
~/data/RNAseqCorrected$ wc -l SRR4030253_2.cor.fq
**31358435** SRR4030253_2.cor.fq

~/data/RNAseqCorrected$ wc -l SRR4030263_1.cor.fq
**57441323** SRR4030263_1.cor.fq
~/data/RNAseqCorrected$ wc -l SRR4030263_2.cor.fq
**57441336** SRR4030263_2.cor.fq

~/data/RNAseqCorrected$ wc -l SRR4030264_1.cor.fq
**7448381** SRR4030264_1.cor.fq
~/data/RNAseqCorrected$ wc -l SRR4030264_2.cor.fq
**7448379** SRR4030264_2.cor.fq
(R1 and R2 of sample SRR4030264 are different in only 2 reads!?)

In general, they differ in "not many" number of reads, anyway, I don't know what is going on (why after using Rcorrector the number of reads between R1 and R2 are different (??)). I viewed some lines of both .fastq and cor.fq files and they look correct.

On the other hand, it seems that not because of having different number of reads in R1 and R2 I get the StopIteration error message because the program worked fine for samples SRR4030252 and SRR4030263, whose R1 and R2 files contain different number of reads (as showed above). I did get the StopIteration with the other two samples that I showed above: SRR4030253 and SRR4030264.

In short, 
1) It seems that the fact that .cor.fq files have different number of R1 and R2 is not affecting the run of FilterUncorrectabledPEfastq.py (as for SRR4030252 and SRR4030263).
2) I still don't know what is going on. I am considering whether I could do something to fix this (once I understand what is the problem) or carry on with the samples that did worked when using FilterUncorrectabledPEfastq.py, but I don't want to do this because only 6 out of 17 samples did not give the StopIteration problem! :S

I would be grateful if you can give me a hint on what could be the problem and how to fix it.

Thanks again, :)

Regards,

'Angela

**Comment by angelaparodymerino** [comment]
* Well, I counted lines, no reads, but anyway R1 and R2 should have same number of lines, if I am right...

**Comment by angelaparodymerino** [comment]
and I think the problem is in Rcorrector: [https://github.com/mourisl/Rcorrector/issues/12 ](url)
I will let you know if I manage to fix this.

**Comment by adamfreedman** [comment]
updated to handle some issues wrt iteration over input files, and removed the "cor" string from the output files ... and added gzipping of output files. code tested (and works) with python 3.10.12

**Comment by brant-ac** [comment]
Hi all,
I am running the FilterUncorrectabledPEfastq.py and I am not sure if the unfixrm...1.cor.fq and unfixrm...2.cor.fq are the new fixed files to be used in the next step, or if they are just the bad quality filtered-out read.
